iceoryx_examples/ice_multi_publisher/ice_re_subscriber.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2020 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/popo/guard_condition.hpp"
+#include "iceoryx_posh/popo/modern_api/typed_subscriber.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "topic_data.hpp"
+
+#include <chrono>
+#include <csignal>
+#include <iostream>
+
+bool killswitch = false;
+
+static void sigHandler(int sig [[gnu::unused]])
+{
+    killswitch = true;
+}
+
+constexpr uint64_t MAX_HISTORY_SIZE{4};
+constexpr uint64_t UNSUBSCRIBED_TIME_SECONDS{3};
+
+void receive()
+{
+    iox::popo::TypedSubscriber<CounterTopic> subscriber({"CounterTopic", iox::capro::AnyInstanceString, "Counter"});
+
+    subscriber.subscribe();
+    uint64_t maxNumSamples = MAX_HISTORY_SIZE - 2;
+    while (!killswitch)
+    {
+        // unsubscribe and resubscribe
+        subscriber.unsubscribe();
+        std::cout << "Unsubscribed ... Resubscribe in " << UNSUBSCRIBED_TIME_SECONDS << " seconds" << std::endl;
+
+        // we will probably miss some data while unsubscribed
+        std::this_thread::sleep_for(std::chrono::seconds(UNSUBSCRIBED_TIME_SECONDS));
+
+        // we (re)subscribe with differing maximum number of samples
+        // and should see at most the latest last maxNumSamples
+        maxNumSamples = maxNumSamples % MAX_HISTORY_SIZE + 1; // cycles between last 1 to MAX_HISTORY_SIZE samples
+        subscriber.subscribe(maxNumSamples);
+
+        std::cout << "Subscribe with max number of samples " << maxNumSamples << std::endl;
+
+        std::this_thread::sleep_for(std::chrono::seconds(1));
+
+        while (subscriber.hasNewSamples())
+        {
+            subscriber.take()
+                .and_then([](iox::popo::Sample<const CounterTopic>& sample) {
```

Comment:
FYI I think you can use auto in the lambda now.
i.e. `.and_then([](auto& sample) {})`

Reply:
you are right, C++14 allows that.

Reply:
Actually, it is not quite as easy with type inference in play like this due to overloads of and_then I think... I will look at the reason, but the call ``*sample.get()`` will not work anymore with auto since it is now detected to be an optional of sample, which is not what it should be (incorrect monadic behavior in my book). Now, changing the call to account for this seems to not play nicely with auto and it breaks. 

Will need a closer look.
 Message: the nested optionals samples and optionals and different overloads for and_then (some incorrect I think) are not working nicely to create an easy user experience.

Reply:
I put that on the list in #408.

Reply:
I suspect we need to change something for sample or the and_then overloads in expected. The type inference of C++ seems to get confused the way it is, but maybe I missed something. It seems a bit tricky but doable, but also minor so it is postponed.